File: src/main/java/com/example/reactive/chapter1/callback/CallbackShoppingCartService.java
package com.example.reactive.chapter1.callback;

import java.util.function.Consumer;

import com.example.reactive.chapter1.commons.Input;
import com.example.reactive.chapter1.commons.Output;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@RequiredArgsConstructor
public class CallbackShoppingCartService implements ShoppingCartService {

    @Override
    public void calculate(Input input, Consumer<Output> consumer) {

        log.info("Callback Style calculating...");

        new Thread(() -> {
            try {
                Thread.sleep(2000);
            } catch (InterruptedException e) {
                log.error(e.getMessage(), e);
            }
            consumer.accept(new Output());
        }).start();
    }
}

File: src/main/kotlin/com/example/reactor/store/callback/CallbackShoppingCartService.kt
package com.example.reactor.store.callback

import com.example.reactor.shared.logger
import com.example.reactor.store.domain.Input

class CallbackShoppingCartService {
    fun calculate(input: Input, runnable: Runnable) {
        Thread {
            Thread.sleep(2_000)
            logger().info("calculated!")
            runnable.run()
        }.start()
    }
}
